pkg/client/object.go

```diff
@@ -0,0 +1,881 @@
+	"github.com/nspcc-dev/neofs-api-go/v2/client"
+	v2object "github.com/nspcc-dev/neofs-api-go/v2/object"
+	v2refs "github.com/nspcc-dev/neofs-api-go/v2/refs"
+	"github.com/nspcc-dev/neofs-api-go/v2/signature"
+	"github.com/pkg/errors"
+)
+
+type Range struct {
+	ln, off uint64
+}
+
+type PutObjectParams struct {
+	obj *object.Object
+
+	r io.Reader
+}
+
+type DeleteObjectParams struct {
+	addr *object.Address
+}
+
+type GetObjectParams struct {
+	addr *object.Address
+
+	w io.Writer
+}
+
+type ObjectHeaderParams struct {
+	addr *object.Address
+
+	short bool
+}
+
+type RangeDataParams struct {
+	addr *object.Address
+
+	r *Range
+
+	w io.Writer
+}
+
+type RangeChecksumParams struct {
+	typ checksumType
+
+	addr *object.Address
+
+	rs []*Range
+
+	salt []byte
+}
+
+type putObjectV2Writer struct {
+	key *ecdsa.PrivateKey
+
+	chunkPart *v2object.PutObjectPartChunk
+
+	req *v2object.PutRequest
+
+	stream v2object.PutObjectStreamer
+}
+
+type checksumType int
+
+const (
+	_ checksumType = iota
+	checksumSHA256
+	checksumTZ
+)
+
+const chunkSize = 3 * (1 << 20)
+
+const TZSize = 64
+
+func rangesToV2(rs []*Range) []*v2object.Range {
+	r2 := make([]*v2object.Range, 0, len(rs))
+
+	for i := range rs {
+		r2 = append(r2, rs[i].toV2())
+	}
+
+	return r2
+}
+
+func (t checksumType) toV2() v2refs.ChecksumType {
+	switch t {
+	case checksumSHA256:
+		return v2refs.SHA256
+	case checksumTZ:
+		return v2refs.TillichZemor
+	default:
+		panic(fmt.Sprintf("invalid checksum type %d", t))
+	}
+}
+
+func (r *Range) WithLength(v uint64) *Range {
+	if r != nil {
+		r.ln = v
+	}
+
+	return r
+}
+
+func (r *Range) WithOffset(v uint64) *Range {
+	if r != nil {
+		r.off = v
+	}
+
+	return r
+}
+
+func (r *Range) toV2() *v2object.Range {
+	if r != nil {
+		r2 := new(v2object.Range)
+		r2.SetOffset(r.off)
+		r2.SetLength(r.ln)
+
+		return r2
+	}
+
+	return nil
+}
+
+func (w *putObjectV2Writer) Write(p []byte) (int, error) {
+	w.chunkPart.SetChunk(p)
+
+	w.req.SetVerificationHeader(nil)
+
+	if err := signature.SignServiceMessage(w.key, w.req); err != nil {
+		return 0, errors.Wrap(err, "could not sign chunk request message")
+	}
+
+	if err := w.stream.Send(w.req); err != nil {
+		return 0, errors.Wrap(err, "could not send chunk request message")
+	}
+
+	return len(p), nil
+}
+
+func (p *PutObjectParams) WithObject(v *object.Object) *PutObjectParams {
+	if p != nil {
+		p.obj = v
+	}
+
+	return p
+}
+
+func (p *PutObjectParams) WithPayloadReader(v io.Reader) *PutObjectParams {
+	if p != nil {
+		p.r = v
+	}
+
+	return p
+}
+
+func (c *Client) PutObject(ctx context.Context, p *PutObjectParams, opts ...CallOption) (*object.ID, error) {
+	// check remote node version
+	switch c.remoteNode.Version.Major {
+	case 2:
+		return c.putObjectV2(ctx, p, opts...)
+	default:
+		return nil, unsupportedProtocolErr
+	}
+}
+
+func (c *Client) putObjectV2(ctx context.Context, p *PutObjectParams, opts ...CallOption) (*object.ID, error) {
+	// convert object to V2
+	obj, err := p.obj.ToV2(c.key)
```

Comment:
Does `p.obj` have set up all available fields at this stage?

Reply:
`ID` and `Signature` will be set by `ToV2` method if needed.

Reply:
So, `ToV2` _completes_ object structure?

Reply:
Yes, it completes structure once. If passed `Object` is already completed, `ToV2` does not change it.